website/project/model.py

```diff
@@ -3563,3 +3572,159 @@ def _on_reject(self, user, token):
+        # TODO: uncomment to set flags
+        #     if not self.registration_schema:
+        #         flags = schema.get('flags', {})
+        #         for flag, value in flags.iteritems():
+        #             self.flags[flag] = value
+
+    # TODO: uncomment to expose approval/review properties
+    #    @property
+    #    def is_pending_review(self):
+    #        return self.flags.get('isPendingReview')
+    #
+    #    @is_pending_review.setter
+    #    def is_pending_review(self, value):
+    #        self.flags['isPendingReview'] = value
+    #
+    #    @property
+    #    def is_approved(self):
+    #        self.flags.get('isApproved')
+    #
+    #    @is_approved.setter
+    #    def is_approved(self, value):
+    #        self.flags['isApproved'] = value
+
+    def update_metadata(self, metadata, save=True):
+        # TODO: uncommnet to disallow editing drafts that are approved or pending approval
+        # if self.is_pending_review or self.is_approved:
+        #    raise NodeStateError('Cannot edit while this draft is being reviewed')
+        changes = []
+        for key, value in metadata.iteritems():
+            old_value = self.registration_metadata.get(key)
+            if not old_value or old_value.get('value') != value.get('value'):
+                changes.append(key)
+        self.registration_metadata.update(metadata)
+        if save:
+            self.save()
+        # TODO: uncomment to nullify approval state if edited
+        # project_signals.draft_edited.send(self, changes)
+        # self.after_edit(changes)
+        return changes
+
+    def before_edit(self, auth):
+        messages = []
+        # TODO: uncomment to provide pre-edit warnings for drafts that are approved or are pending approval
+        # if self.flags.get('isApproved'):
+        #     messages.append('The draft registration you are editing is currently approved. Please note that if you make any changes (excluding comments) this approval status will be revoked and you will need to submit for approval again.')
+        # if self.flags.get('isPendingReview'):
+        #     messages.append('The draft registration you are editing is currently pending review. Please note that if you make any changes (excluding comments) this request will be cancelled and you will need to submit for approval again.')
+        return messages
+
+    def after_edit(self, changes):
+        pass
+        # TODO: uncomment when we support review/approval
+        ## revoke approval and review status if changed
+        #if changes:
+        #    self.flags.update({
+        #        'isPendingReview': False,
+        #        'isApproved': False
+        #    })
+        #    self.save()
+
+    def find_question(self, qid):
+        for page in self.registration_schema.schema['pages']:
+            for question_id, question in page['questions'].iteritems():
+                if question_id == qid:
+                    return question
```

Comment:
This is an extremely inefficient way to find a given question; do we have any flexibility over the data structure for schemas? If so, this is a prime candidate for refactoring.

Sorting schemas by page number seems particularly incongruous given that the demo I saw only showed one question at a time.